### ISSUE TITLE
rustfmt hangs on long method chain

### ISSUE THREAD

**Issue by tomprince** [rust-lang-nursery/rustfmt#1265]
`rustfmt` runs for over two minutes without completing on the following file.

https://github.com/kstep/rust-mpd/blob/a43b0ffa8ffbd5bc97d489ef3526f2ab17d590c5/src/status.rs

**Comment by nrc** [comment]
It does seem to be the long method chain which causes the hang. Here is the relevant part of the code:

```
fn encode<S: Encoder>(e: &mut S) -> Result<(), S::Error> {
    e.emit_struct("Status", 21, |e| {
        e.emit_struct_field("volume", 0, |e| self.volume.encode(e))
         .and_then(|_| e.emit_struct_field("repeat", 1, |e| self.repeat.encode(e)))
         .and_then(|_| e.emit_struct_field("random", 2, |e| self.random.encode(e)))
         .and_then(|_| e.emit_struct_field("single", 3, |e| self.single.encode(e)))
         .and_then(|_| e.emit_struct_field("consume", 4, |e| self.consume.encode(e)))
         .and_then(|_| e.emit_struct_field("queue_version", 5, |e| self.queue_version.encode(e)))
         .and_then(|_| e.emit_struct_field("queue_len", 6, |e| self.queue_len.encode(e)))
         .and_then(|_| e.emit_struct_field("state", 7, |e| self.state.encode(e)))
         .and_then(|_| e.emit_struct_field("song", 8, |e| self.song.encode(e)))
         .and_then(|_| e.emit_struct_field("nextsong", 9, |e| self.nextsong.encode(e)))
         .and_then(|_| {
             e.emit_struct_field("time", 10, |e| {
                 e.emit_option(|e| {
                     self.time
                         .map(|p| {
                             e.emit_option_some(|e| {
                                 e.emit_tuple(2, |e| {
                                     e.emit_tuple_arg(0, |e| p.0.num_seconds().encode(e))
                                      .and_then(|_| e.emit_tuple_arg(1, |e| p.1.num_seconds().encode(e)))
                                 })
                             })
                         })
                         .unwrap_or_else(|| e.emit_option_none())
                 })
             })
         })
         .and_then(|_| {
             e.emit_struct_field("elapsed", 11, |e| {
                 e.emit_option(|e| {
                     self.elapsed
                         .map(|d| e.emit_option_some(|e| d.num_seconds().encode(e)))
                         .unwrap_or_else(|| e.emit_option_none())
                 })
             })
         })
         .and_then(|_| {
             e.emit_struct_field("duration", 12, |e| {
                 e.emit_option(|e| {
                     self.duration
                         .map(|d| e.emit_option_some(|e| d.num_seconds().encode(e)))
                         .unwrap_or_else(|| e.emit_option_none())
                 })
             })
         })
         .and_then(|_| e.emit_struct_field("bitrate", 13, |e| self.bitrate.encode(e)))
         .and_then(|_| {
             e.emit_struct_field("crossfade", 14, |e| {
                 e.emit_option(|e| {
                     self.crossfade
                         .map(|d| e.emit_option_some(|e| d.num_seconds().encode(e)))
                         .unwrap_or_else(|| e.emit_option_none())
                 })
             })
         })
         .and_then(|_| e.emit_struct_field("mixrampdb", 15, |e| self.mixrampdb.encode(e)))
         .and_then(|_| {
             e.emit_struct_field("mixrampdelay", 16, |e| {
                 e.emit_option(|e| {
                     self.mixrampdelay
                         .map(|d| e.emit_option_some(|e| d.num_seconds().encode(e)))
                         .unwrap_or_else(|| e.emit_option_none())
                 })
             })
         })
         .and_then(|_| e.emit_struct_field("audio", 17, |e| self.audio.encode(e)))
         .and_then(|_| e.emit_struct_field("updating_db", 18, |e| self.updating_db.encode(e)))
         .and_then(|_| e.emit_struct_field("error", 19, |e| self.error.encode(e)))
         .and_then(|_| e.emit_struct_field("replaygain", 20, |e| self.replaygain.encode(e)))
    })
}
```

That's a method call chain 21 deep. I suspect we have some O(n^2) or worse behaviour going on here, and it feels like we really shouldn't need to. Once we're in to multi-line mode, we should not keep trying to do better.

**Comment by koivunej** [comment]
I tried to do some digging around this. My issue is with code generated by `quick-protobuf`, where generated structs each get methods like:

```
fn get_size(&self) -> usize {
  0
  + self.first_field.as_ref().map_or(0, |m| 1 + sizeof_varint(*(m) as u64))
  + self.second_field.
  + ...
  ...
}
```

I first did some profiling with `prof` using `torch.sh` but it revealed very little, pointing out `comment::recover_comment_removed` as the most suspicious. It can be made to do an order of magnitude less work (at least in debug builds) by checking first if the `snippet` contains any comment characters using regex, but that does not help with these long expressions.

What I gather happens with these long chained statements/expressions is that after processing each part of sum in the case of `quick-protobuf` the whole block is reprocessed by descending from [`Rewrite` trait impl for `ast::Block` before returning](https://github.com/rust-lang-nursery/rustfmt/blob/master/src/expr.rs#L627). If I am correct, this could explain the hours of running time. As an another kind of confirmation, I tried rewriting code produced by `quick-protobuf` code as follows makes `rustfmt` run complete quickly: 

```
let mut sum = 0;
sum += ...;
sum += ...;
```

No idea how to fix this. Simply removing the descending did not work :) 

Perhaps as a more important side-note it'd appear that `#[cfg_attr(rustfmt, rustfmt_skip)]` does *not* stop processing of the module, as it probably should given how long these kinds of inputs can take to go through with the current design. Should a new bug be filed on that?

**Comment by koivunej** [comment]
Doing some `time rustfmt` runs I found out that the "`snippet` contains any comment characters using regex" hack does cause quite a speedup with [rust_types.rs](https://github.com/tafia/quick-protobuf/blob/master/examples/codegen/data_types.rs) as input. I've published the patch and timings as a [gist](https://gist.github.com/koivunej/2f6dd73508cd9bca29403175ebc71817).

In short: running time goes from 75s to 50s. That does not solve this issue though, but it might be something to consider later on.

**Comment by davidalber** [comment]
# tl;dr
I'm not able to repro this. Formatting the file takes less than one second on my machine. Perhaps it has been fixed in the intervening time.

# Methodology
I created two files:
- 1265.rs contains all of [the original file](https://github.com/kstep/rust-mpd/blob/a43b0ffa8ffbd5bc97d489ef3526f2ab17d590c5/src/status.rs).
- 1265-2.rs contains the snippet from @nrc's [comment](https://github.com/rust-lang-nursery/rustfmt/issues/1265#issuecomment-275501831) wrapped in `impl Encodable for Status { }`.

Here's my rustfmt.toml:
```toml
error_on_line_overflow_comments = false
```

Without that attribute, formatting 1265.rs failed with the following message.
```
error: line exceeded maximum width (maximum: 100, found: 139)
```

## Version
```
$ rustfmt --version
0.2.15-nightly ( )
```

## Formatting Full File
```
$ time rustfmt --write-mode=display -- 1265.rs > /dev/null

real    0m0.745s
user    0m0.704s
sys     0m0.031s
```

## Formatting Snippet
```
$ time rustfmt --write-mode=display -- 1265-2.rs > /dev/null

real    0m0.744s
user    0m0.706s
sys     0m0.032s
```

With the redirection to /dev/null removed, rustfmt returns formatted code.

**Comment by nrc** [comment]
Thanks for investigating @davidalber !
